dbt/adapters/bigquery/impl.py

```diff
@@ -709,6 +710,35 @@ def upload_file(self, local_file_path: str, database: str, table_schema: str,
         with self.connections.exception_handler("LOAD TABLE"):
             self.poll_until_job_completes(job, timeout)
 
+    @available.parse_none
+    def upload_run_results(self, local_file_path: str, database: str, table_schema: str,
```

Comment:
I'd like to see this method be more generic in one way, _and_ more narrowly targeted in another.
- More generic: Could we rename to just `upload_json_artifact`? Is there a reason why this only works for `run_results.json`, and wouldn't work for `manifest.json`, `catalog.json`, or `sources.json`?
- More narrowly targeted: The intent with this method is _not_ to support arbitrary file &rarr; table creation. It's very specifically to hack support for uploading metadata artifacts, and I'd like to keep it as narrowly targeted as possible. I'm not sure of the best way to do that—perhaps we could validate that the local file is named `manifest.json`, `catalog.json`, `run_results.json`, or `sources.json`?

To that end, I'd also propose we also make the even-more-generic `upload_file` _unavailable_ to the Jinja context — i.e. remove the `@available.parse_none`, and remove the `bigquery__upload_file` macro. (It's still good to have the separate method, as a way of factoring the code here.)

Reply:
Thanks for the input @jeremy. Some thoughts on the latest commits I've pushed:
* Furthering your thoughts, if the desire is for a macro that uploads json artifacts and provides no other functionality then the file name selection is redundant. Instead a directory can be specified with the artifact file names hard-coded on the `dbt-bigquery` side. Does this make sense on your side, is there any way to create artifacts with custom names that I'm not aware of?
* BigQuery does not allow "." or "-" characters in column names. By default these characters will be replaced with double underscores ("__") although the option to select your own replacement character(s) is provided.
* BigQuery does not like lists inside lists as present with the "ref" key in `manifest.json`. For example, this file cannot be uploaded to Bigquery:
    ```
    {"col1": 1, "col2": [[1, 2]]}
    ```
    As it returns the error message "Nested arrays not allowed". In contrast this file can be uploaded:
    ```
    {"col1": 1, "col2": [1, 2]}
    ```
    The `upload_json_artifacts` method handles this for `manifest.json` by replacing the outer list with the value of the inner list. I'm not familiar with the values that can be contained in the "ref" key, do you see any issue with this approach?